font/src/directwrite/mod.rs

```diff
@@ -122,15 +124,20 @@ impl crate::Rasterize for DirectWriteRasterizer {
 
         let offset = GlyphOffset { advanceOffset: 0.0, ascenderOffset: 0.0 };
 
-        let glyph_index = *font
-            .get_glyph_indices(&[glyph.c as u32])
-            .first()
-            .ok_or_else(|| Error::MissingGlyph(glyph.c))?;
-        if glyph_index == 0 {
-            // The DirectWrite documentation states that we should get 0 returned if the glyph
-            // does not exist in the font
-            return Err(Error::MissingGlyph(glyph.c));
-        }
+        let glyph_index: u16 = match glyph.id {
+            KeyType::GlyphIndex(i) => i as u16,
+            // On placeholder return a missing glyph error with the placeholder unicode character
+            KeyType::Placeholder => *font
+                .get_glyph_indices(&[' ' as u32])
+                .first()
+                .filter(|index| **index != 0)
+                .ok_or_else(|| Error::MissingGlyph(' '))?,
+            KeyType::Char(c) => *font
+                .get_glyph_indices(&[c as u32])
+                .first()
+                .filter(|index| **index != 0)
+                .ok_or_else(|| Error::MissingGlyph(c))?,
+        };
```

Comment:
Just had an idea. I think we can do things a lot better here if we never actually receive GlyphIndex.

In this match arm we can just return either `c` or `' '` and mark the `GlyphIndex` branch an usreachable. Then we do all the other stuff once instead of twice, since the code is identical.

This allows us to get rid of `keytype_unwrap_char` too!